File: PHP/petooh.php
<?php

  class Interpreter {

    const OP_INC = "Ko";
    const OP_DEC = "kO";
    const OP_INCPTR = "Kudah";
    const OP_DECPTR = "kudah";
    const OP_OUT = "Kukarek";
    const OP_JMP = "Kud";
    const OP_RET = "kud";
    const VALID_CHARS = "adehkKoOru";

    private $cells = [];
    private $cellptr = 0;
    private $stack = [];
    private $level = 0;

    private function _push ($instruction) {
      $this->stack[$this->level][] = $instruction;
    }

    private function _loop () {
      while ($this->cells[$this->cellptr] > 0)
        foreach ($this->stack[$this->level] as $instruction)  {
        if (!array_key_exists($this->cellptr, $this->cells))
          $this->cells[$this->cellptr] = 0;
        if ($instruction == self::OP_INC) $this->cells[$this->cellptr] += 1;
        elseif ($instruction == self::OP_DEC) $this->cells[$this->cellptr] -= 1;
        elseif ($instruction == self::OP_INCPTR) $this->cellptr += 1;
        elseif ($instruction == self::OP_DECPTR) $this->cellptr -= 1;
        elseif ($instruction == self::OP_OUT) print(chr($this->cells[$this->cellptr]));
      }
    }

    public function evaluate ($code) {
      $code = str_split(preg_replace("/[^" . self::VALID_CHARS . "]*/", '', $code));
      $buffer = '';

      foreach ($code as $char) {
        $instruction = $buffer . $char;

        if (!array_key_exists($this->cellptr, $this->cells))
          $this->cells[$this->cellptr] = 0;

        if ($instruction == self::OP_INC) {
          $buffer = '';
          if ($this->level > 0) 
            $this->_push($instruction);
          else 
            $this->cells[$this->cellptr] += 1;
        } elseif ($instruction == self::OP_DEC) {
            $buffer = '';
            if ($this->level > 0)
                $this->_push($instruction);
            else
                $this->cells[$this->cellptr] -= 1;
        } elseif ($instruction == self::OP_INCPTR) {
            $buffer = '';
            if ($this->level > 0)
                $this->_push($instruction);
            else
                $this->cellptr += 1;
        } elseif ($instruction == self::OP_DECPTR) {
            $buffer = '';
            if ($this->level > 0) 
                $this->_push($instruction);
            else
                $this->cellptr -= 1;
        } elseif ($buffer == self::OP_JMP && $char != 'a') {
            $buffer = $char;
            $this->level += 1;
            $this->stack[$this->level] = [];
        } elseif ($buffer == self::OP_RET && $char != 'a') {
            $buffer = $char;
            $this->_loop();
            $this->level -= 1;
        } elseif ($instruction == self::OP_OUT) {
            $buffer = '';
            if ($this->level > 0)
                $this->_push($instruction);
            else
                print(chr($this->cells[$this->cellptr]));
        } else
            $buffer .= $char;
      }
    }

  }

  if (count($argv) == 2) {
    try {
      $interpreter = new Interpreter();
      $interpreter->evaluate(file_get_contents($argv[1]));
    }
    catch (Exception $e) {
      print $e;
    }
  } else {
    printf("\n    USAGE: php %s /path/to/file.koko\r\n", $argv[0]);
  }

File: Python/petooh.py
import sys
from collections import defaultdict


class Interpreter(object):

    OP_INC = "Ko"
    OP_DEC = "kO"
    OP_INCPTR = "Kudah"
    OP_DECPTR = "kudah"
    OP_OUT = "Kukarek"
    OP_JMP = "Kud"
    OP_RET = "kud"
    VALID_CHARS = "adehkKoOru"
    MAX_MEM = 65535

    def __init__(self, maxmem=MAX_MEM):
        self.cells, self.cellptr = [0]*maxmem, 0
        self.stack, self.level = defaultdict(list), 0

    def _push(self, instruction):
        self.stack[self.level].append(instruction)

    def _loop(self):
        while self.cells[self.cellptr] > 0:
            for instruction in self.stack[self.level]:
                if instruction == self.OP_INC: self.cells[self.cellptr] += 1
                elif instruction == self.OP_DEC: self.cells[self.cellptr] -= 1
                elif instruction == self.OP_INCPTR: self.cellptr += 1
                elif instruction == self.OP_DECPTR: self.cellptr -= 1
                elif instruction == self.OP_OUT: sys.stdout.write(chr(self.cells[self.cellptr]))

    def evaluate(self, code):
        code = filter(self.VALID_CHARS.__contains__, code)
        buffer = ''

        try:
            for char in code:
                instruction = buffer + char

                if instruction == self.OP_INC:
                    buffer = ''
                    if self.level > 0:
                        self._push(instruction)
                    else:
                        self.cells[self.cellptr] += 1
                elif instruction == self.OP_DEC:
                    buffer = ''
                    if self.level > 0:
                        self._push(instruction)
                    else:
                        self.cells[self.cellptr] -= 1
                elif instruction == self.OP_INCPTR:
                    buffer = ''
                    if self.level > 0:
                        self._push(instruction)
                    else:
                        self.cellptr += 1
                elif instruction == self.OP_DECPTR:
                    buffer = ''
                    if self.level > 0:
                        self._push(instruction)
                    else:
                        self.cellptr -= 1
                elif buffer == self.OP_JMP and char != 'a':
                    buffer = char
                    self.level += 1
                    self.stack[self.level] = []
                elif buffer == self.OP_RET and char != 'a':
                    buffer = char
                    self._loop()
                    self.level -= 1
                elif instruction == self.OP_OUT:
                    buffer = ''
                    if self.level > 0:
                        self._push(instruction)
                    else:
                        sys.stdout.write(chr(self.cells[self.cellptr]))
                else:
                    buffer += char
        except IndexError:
            print "[Error] Segmentation fault at address 0x%08x (maxmem = %s)" % (self.cellptr, len(self.cells))


if __name__ == "__main__":
    if len(sys.argv) == 2:
        try:
            with open(sys.argv[1], "r") as f:
                interpreter = Interpreter()
                interpreter.evaluate(f.read())
        except IOError as e:
            print e
    else:
        print "\n    USAGE: %s /path/to/file.koko" % sys.argv[0]